include/net/downloader.h

```diff
@@ -0,0 +1,358 @@
+
+	/** Ensure that thread is ready for download */
+	struct k_sem event_sem;
+	/** Protect shared variables. */
+	struct k_mutex mutex;
+	/** Downloader state. */
+	enum downloader_state state;
+	/** Internal download thread. */
+	struct k_thread thread;
+	/** Internal thread ID. */
+	k_tid_t tid;
+	/* Internal thread stack. */
+	K_KERNEL_STACK_MEMBER(thread_stack, CONFIG_DOWNLOADER_STACK_SIZE);
+};
+
+/**
+ * @brief Initialize the downloader.
+ *
+ * @param[in] dl	Downloader instance.
+ * @param[in] config	Downloader configuration options.
+ *
+ * @return Zero on success, otherwise a negative error code.
+ */
+int downloader_init(struct downloader *dl, struct downloader_cfg *config);
+
+/**
+ * @brief Deinitialize the downloader.
+ *
+ * @param[in] dl Downloader instance.
+ *
+ * @return Zero on success.
+ */
+int downloader_deinit(struct downloader *dl);
+
+/**
+ * @brief Download a file asynchronously.
+ *
+ * This initiates an asynchronous connect-download-disconnect sequence to the target
+ * host.
+ *
+ * Downloads are handled one at a time. If previous download is not finished
+ * this returns -EALREADY.
+ *
+ * @param[in] dl		Downloader instance.
+ * @param[in] host_config	Host configuration options.
+ * @param[in] url		URI of the host to connect to.
+ *				Can include scheme, port number and full file path, defaults to
+ *				HTTP or HTTPS if no scheme is provided.
+ * @param[in] from		Offset from where to resume the download,
+ *				or zero to download from the beginning.
+ *
+ * @return Zero on success, a negative error code otherwise.
+ */
+int downloader_get(struct downloader *dl, const struct downloader_host_cfg *host_config,
+		   const char *url, size_t from);
+
+/**
+ * @brief Download a file asynchronously with host and file as separate parameters.
+ *
+ * This initiates an asynchronous connect-download-disconnect sequence to the target
+ * host.
+ *
+ * Downloads are handled one at a time. If previous download is not finished,
+ * this returns -EALREADY.
+ *
+ * @param[in] dl		Downloader instance.
+ * @param[in] host_config	Host configuration options.
+ * @param[in] host		URI of the host to connect to.
+ *				Can include scheme and port number, defaults to
+ *				HTTP or HTTPS if no scheme is provided.
+ * @param[in] file		URL to download.
+ * @param[in] from		Offset from where to resume the download,
+ *				or zero to download from the beginning.
+ *
+ * @return Zero on success, a negative error code otherwise.
+ */
+int downloader_get_with_host_and_file(struct downloader *dl,
+				      const struct downloader_host_cfg *host_config,
+				      const char *host, const char *file, size_t from);
+
+/**
+ * @brief Cancel file download.
+ *
+ * Request downloader to stop the download. This does not block.
+ * If the @c downloader_host_cfg.keep_connection flag is set the downloader remains connected to
+ * the server. Else the downloader is disconnected and the @c DOWNLOADER_EVT_STOPPED event is sent.
```

Comment:
If the `keep_connection` flag is set and `downloader_cancel()` is called, is there a `DOWNLOADER_EVT_STOPPED` event or not? Based on the description I get the impression that it would only be sent when `keep_connection` is false, but I think it would make sense that the event is sent in both cases, indicating that the library is ready for a new download.

It seems that the `DOWNLOADER_EVT_STOPPED` event has its roots in `DOWNLOAD_CLIENT_EVT_CLOSED`. This explains where the unclarity comes from. If the new event is only sent when the socket is closed, then it should probably be named differently (like it's named in the Download client). If it's sent sent always when the download has stopped, then the documentation needs to be updated.

Reply:
It is sent in both cases. I have updated the documentation to make it more clear that it is sent on all calls to _cancel().